dashboard/pages/daily_report.py

```diff
@@ -0,0 +1,98 @@
+            host=environ['DB_HOST'],
+            dbname=environ['DB_NAME'],
+            user=environ['DB_USERNAME'],
+            password=environ['DB_PASSWORD'],
+            port=environ['DB_PORT']
+        )
+    except Exception as e: # pylint: disable=broad-exception-caught
+        logging.error(e)
+        return None
+
+
+def get_db_cursor(conn: connection) -> cursor | None:
+    """return a cursor object based on a given connection"""
+    try:
+        return conn.cursor(cursor_factory=RealDictCursor)
+    except Exception as e: # pylint: disable=broad-exception-caught
+        logging.error(e)
+        return None
+
+
+def is_email_already_subscribed(email: str, conn: connection) -> bool:
+    """check if the email given has already subscribed to the database"""
+    with get_db_cursor(conn) as curs:
+        curs.execute("SELECT COUNT(*) FROM subscriber WHERE email=%s", (email, ))
+        res = curs.fetchone()
+    return res["count"] == 1
+
+
+def is_email_valid(email: str) -> bool:
+    """return a bool based on whether this is a valid email"""
+    email_pattern = re.compile(r"[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}")
+    return re.fullmatch(pattern=email_pattern, string=email) is not None
+
+
+def get_email_from_subscription_form() -> str | None:
+    """create a subscription form to the summary report"""
+    with st.form("summary_subscription", clear_on_submit=True, border=True):
+        email = st.text_input("Email:")
+        submit = st.form_submit_button()
+
+    if email and submit:
+        if is_email_valid(email):
+            return email
+        st.error("This is not a valid email.")
+    return None
+
+def upload_new_subscriber(conn: connection, email: str):
+    """Take the email given and attempt to upload it to the database. """
+    try:
+        with get_db_cursor(conn) as curs:
+            curs.execute("INSERT INTO subscriber (email) VALUES (%s)", (email, ))
+            conn.commit()
+        return 1
+    except Exception as e: # pylint: disable=broad-exception-caught
+        logging.error(e)
+        return None
+
+def deploy_subscription_page():
+    """This contains the main code for the subscription page."""
+    st.title("Summary Report")
+    st.subheader(
+        "Subscribe to a daily report on our tracking results by adding your email below.")
+    subscriber_email = get_email_from_subscription_form()
+    conn = get_db_connection()
+    if subscriber_email and is_email_already_subscribed(subscriber_email, conn):
+        st.error("This email is already subscribed to the summary report.")
+    elif subscriber_email:
+        if upload_new_subscriber(conn, subscriber_email) is not None:
+            st.success(f"{subscriber_email} is now subscribed to the summary report.")
+            logging.info(f"New email added: {subscriber_email}")
+        else:
+            st.error(f"An error ocurred when trying to add {subscriber_email} to the database.")
+
+
+if __name__ == "__main__":
+    load_dotenv()
```

Comment:
I believe that is only the case for lambda functions since they invoke the function rather than run the file as main. Since we're running this as a service I do not think this will cause issues.

Reply:
that sounds very true!